### PR TITLE
NO-SNOW http connection pooling manager

### PR DESCRIPTION
**Pull request by sfc-gh-japatel**
This is being approved earlier, just pushing it to master. 

https://github.com/snowflakedb/snowflake-ingest-java/pull/184

- Added a new commit which fixes the bug saw earlier in fidelity's case. 
- Will also fix the exception in KC once release 0.10.4-beta.3 is released

### CODE REVIEW

**Review by sfc-gh-tjones — APPROVED** [review]

File: src/main/java/net/snowflake/ingest/utils/HttpUtil.java [hunk 1]
```diff
@@ -50,11 +54,39 @@ public class HttpUtil {
...
+  private static IdleConnectionMonitorThread idleConnectionMonitorThread;
+
+  /**
+   * This lock is to synchronize on idleConnectionMonitorThread to avoid setting starting a thread
+   * which was already started. (To avoid {@link IllegalThreadStateException})
+   */
+  private static ReentrantLock idleConnectionMonitorThreadLock = new ReentrantLock(true);
+
+  private static final int DEFAULT_CONNECTION_TIMEOUT_MINUTES = 1;
+  private static final int DEFAULT_HTTP_CLIENT_SOCKET_TIMEOUT_MINUTES = 5;
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/utils/HttpUtil.java:68]: we already have these two above? line 53-54

File: src/main/java/net/snowflake/ingest/utils/HttpUtil.java [hunk 2]
```diff
@@ -126,6 +170,27 @@ private static void initHttpClient() {
...
+      if (connectionManager != null
+          && (idleConnectionMonitorThread == null || idleConnectionMonitorThread.isShutdown())) {
+        // Monitors in a separate thread where it closes any idle connections
+        // https://hc.apache.org/httpcomponents-client-4.5.x/current/tutorial/html/connmgmt.html
+        idleConnectionMonitorThread = new IdleConnectionMonitorThread(connectionManager);
+        idleConnectionMonitorThread.setDaemon(true);
+        idleConnectionMonitorThread.start();
+      }
+    } catch (Exception e) {
+      LOGGER.warn("Unable to start Daemon thread for Http Idle Connection Monitoring", e);
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/utils/HttpUtil.java:188]: logger.error?

File: src/main/java/net/snowflake/ingest/utils/HttpUtil.java [hunk 3]
```diff
@@ -235,4 +300,86 @@ public static Properties generateProxyPropertiesForJDBC() {
...
+            // Close expired connections
+            connectionManager.closeExpiredConnections();
+            // Optionally, close connections
+            // that have been idle longer than 30 sec
+            connectionManager.closeIdleConnections(
+                DEFAULT_IDLE_CONNECTION_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+          }
+        }
+      } catch (InterruptedException ex) {
+        LOGGER.debug("Terminating Idle Connection Monitor Thread ");
```
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/utils/HttpUtil.java:346]: logger.error?

**Review by sfc-gh-tzhang — APPROVED** [review]

File: src/main/java/net/snowflake/ingest/utils/HttpUtil.java (hunk 1 above)
Comment (sfc-gh-japatel) [reply src/main/java/net/snowflake/ingest/utils/HttpUtil.java:68]: Ouch, thanks for catching, let me fix it.

File: src/main/java/net/snowflake/ingest/utils/HttpUtil.java (hunk 2 above)
Comment (sfc-gh-japatel) [reply src/main/java/net/snowflake/ingest/utils/HttpUtil.java:188]: Didnt want to throw this as error, as this is more for debugging. warn should be fine.

File: src/main/java/net/snowflake/ingest/utils/HttpUtil.java (hunk 3 above)
Comment (sfc-gh-japatel) [reply src/main/java/net/snowflake/ingest/utils/HttpUtil.java:346]: Same as before. this is all BG task and is not essential to find the rootcause. will change this to warn.

**Pull request merged**